HPCC-29332 Add opentelemetry based tracing

rpastrana opened the pull request:
- Adds opentel cpp lib + otlp + otlp-http

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

rpastrana committed:
HPCC-29332 Add opentelemetry-cpp to vcpkg list

- Adds opentel cpp lib + otlp + otlp-http

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

rpastrana commented:
@GordonSmith I will add cmake option to suppress this feature if it makes sense in the future, right now it seems open telemetry will be part of the 9.4+ platform.
Also, I added this PR: https://github.com/hpcc-systems/vcpkg/pull/32
let me know if that needs to change

GordonSmith reviewed (changes requested):
@rpastrana you should include this change in your PR for Open Telemetry Support as there is no benefit to having it in master without it (and there is a slight downside to including it when not needed)?

rpastrana committed:
HPCC-29332 Add opentelemetry-cpp to vcpkg list

- WIP

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

rpastrana commented:
> @rpastrana you should include this change in your PR for Open Telemetry Support as there is no benefit to having it in master without it (and there is a slight downside to including it when not needed)?

Thanks @GordonSmith, that makes sense, Gavin's Jira listed this change as a stand-alone subtask but I can merge together.
Can you comment on the actual change, is it fine as-is or does it need any changes before eventual merging?

GordonSmith commented:
That change to vcpkg.json.in looks fine - this is an "always on" thing?

rpastrana commented:
It's likely to be an always-on feature once implemented. What are the vcpkg side effects of making it optional based on a cmake flag?

GordonSmith commented:
You wouldn't make it optional if its always needed - so what you have is good.

rpastrana committed:
Merge branch 'master' of https://github.com/hpcc-systems/HPCC-Platform into HPCC-29332-OpenTelCpp

rpastrana committed:
HPCC-29332 Add opentelemetry-cpp to vcpkg list

- WIP
- Adds system/tracing/tracemanager
- Adds instrumented server/client samples

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

GordonSmith commented on system/tracing/CMakeLists.txt:
```diff
@@ -0,0 +1,73 @@
+################################################################################
+#    HPCC SYSTEMS software Copyright (C) 2012 HPCC Systems®.
```
Year is wrong (do we need this block anyway?)

GordonSmith commented on system/tracing/CMakeLists.txt:
```diff
@@ -0,0 +1,73 @@
+        jlib
+    )
+else()
+    MESSAGE(STATUS "Opentelemetry libs/headers not found")
```
Are these needed - it will fail above with the REQUIRED?

GordonSmith commented on system/tracing/CMakeLists.txt:
```diff
@@ -0,0 +1,73 @@
+         ${CMAKE_BINARY_DIR}/oss
+         ${HPCC_SOURCE_DIR}/system/tracing
+    )
+    #SET ( SRCS tracemanager.hpp opentel/instrumented/httpserver.cpp opentel/instrumented/httpclient.cpp)
```
Commented code.

GordonSmith commented on system/tracing/CMakeLists.txt:
```diff
@@ -0,0 +1,73 @@
+        opentelemetry-cpp::ostream_span_exporter
+        #opentelemetry-cpp::jaeger_trace_exporter
+        #opentelemetry-cpp::otlp_grpc_exporter
+        #opentelemetry-cpp::otlp_grpc_log_exporter
```
Commented code

GordonSmith commented on system/tracing/opentel/instrumented/CMakeLists.txt:
```diff
@@ -0,0 +1,78 @@
+################################################################################
+#    HPCC SYSTEMS software Copyright (C) 2012 HPCC Systems®.
```
Year

GordonSmith commented on system/tracing/CMakeLists.txt:
```diff
@@ -0,0 +1,73 @@
+
+
+
+
```
Blank lines.
Also won't it need an install?

GordonSmith commented on system/tracing/opentel/instrumented/CMakeLists.txt:
```diff
@@ -0,0 +1,78 @@
+# Description:
+# ------------
+#    Cmake Input File for jlib
+#####################################################
```
Bad comments

GordonSmith commented on system/tracing/CMakeLists.txt:
```diff
@@ -0,0 +1,73 @@
+# Description:
+# ------------
+#    Cmake Input File for jlib
+#####################################################
```
Bad comments

GordonSmith commented on system/tracing/opentel/instrumented/CMakeLists.txt:
```diff
@@ -0,0 +1,78 @@
+#For http exporter
+#find_package(CURL REQUIRED)
+
+#find_package(opentelemetry-cpp CONFIG REQUIRED)
```
Commented code.

GordonSmith commented on system/tracing/opentel/instrumented/CMakeLists.txt:
```diff
@@ -0,0 +1,78 @@
+
+#find_package(opentelemetry-cpp CONFIG REQUIRED)
+
+if (opentelemetry-cpp_FOUND)
```
This file won't be included if its not found?

GordonSmith commented on system/tracing/opentel/instrumented/CMakeLists.txt:
```diff
@@ -0,0 +1,78 @@
+#find_package(opentelemetry-cpp CONFIG REQUIRED)
+
+if (opentelemetry-cpp_FOUND)
+    MESSAGE(STATUS "Found Opentelemetry")
```
Not needed

GordonSmith commented on system/tracing/opentel/instrumented/CMakeLists.txt:
```diff
@@ -0,0 +1,78 @@
+        opentelemetry-cpp::ostream_span_exporter
+        #opentelemetry-cpp::jaeger_trace_exporter
+        #opentelemetry-cpp::otlp_grpc_exporter
+        #opentelemetry-cpp::otlp_grpc_log_exporter
```
Commented code

GordonSmith commented on system/tracing/opentel/instrumented/CMakeLists.txt:
```diff
@@ -0,0 +1,78 @@
+        opentelemetry-cpp::http_client_curl
+        jlib)
+else()
+    MESSAGE(STATUS "Opentelemetry libs/headers not found")
```
Not needed

GordonSmith commented on system/tracing/opentel/instrumented/CMakeLists.txt:
```diff
@@ -0,0 +1,78 @@
+        jlib)
+else()
+    MESSAGE(STATUS "Opentelemetry libs/headers not found")
+endif()
```
Should there be an INSTALL?

GordonSmith reviewed (changes requested):
I only looked at the cmake stuff.

rpastrana committed:
HPCC-29332 WIP

- Removes system/tracing/tracemanager

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

rpastrana commented:
@ghalliday this PR is nowhere near ready for review but as discussed, wanted to share with you to review at a very high level while I'm out. We can discuss when I return

ghalliday commented on system/jlib/jtrace.hpp:
```diff
@@ -18,21 +18,197 @@
 #define JTRACE_HPP
 
+#undef UNIMPLEMENTED //opentelemetry defines UNIMPLEMENTED
+#include "opentelemetry/exporters/ostream/span_exporter_factory.h"
```
Avoid these includes in the header file if at all possible...  Otherwise everything in the system is suddenly going to be including many more header files.

ghalliday commented on system/jlib/jtrace.hpp:
```diff
@@ -18,21 +18,197 @@
+namespace opentel_trace = opentelemetry::trace;
+
+//#include "jexcept.hpp" //re-define UNIMPLEMENTED
+#define UNIMPLEMENTED throw makeStringExceptionV(-1, "UNIMPLEMENTED feature at %s(%d)", sanitizeSourceFile(__FILE__), __LINE__)
```
There are cleaner ways of achieving this
#define OLD_X X
#undef X
#include....
#undef X
#define X OLD_X

But it also shouldn't be in the header file unless there is no other solution.

ghalliday commented on system/jlib/jtrace.hpp:
```diff
@@ -18,21 +18,197 @@
+// virtual Keys(nostd::function_ref<bool(nostd::string_view)> /* callback */) 
+//         list of all the keys in the carrier.
+//         By default, it returns true without invoking callback */
+template <typename T>
```
This looks wrong being a template class - it suggests a design decision hasn't been made, and requires everything is in the header - which will cause compile sloath.

rpastrana replied:
This largely as-is example from OpenTel doesn't look wrong to me given what they were hoping to accomplish (get,set,keys for various representations of "httpheaders"), but we can provide certainly go a different way (do I recall you mentioning a lambda function interface )?

ghalliday commented on esp/bindings/http/platform/httpservice.cpp:
```diff
@@ -184,6 +184,34 @@ void checkSetCORSAllowOrigin(EspHttpBinding *binding, CHttpRequest *req, CHttpRe
+
+    //Declare the span, provide appropriate attributes, and options
+    //Trace ID generated if no parent context is provided
+    auto processingRequestSpan =
```
I would expect much of this code to be encapsulated inside jtrace.  That may not be possible, but having open telemetry code that manages spans scattered through the code suggests poor encapsulation.

ghalliday commented on esp/bindings/http/platform/httpservice.cpp:
```diff
@@ -450,6 +483,9 @@ int CEspHttpServer::processRequest()
     }
 
+    //need to ensure that the span is ended when out of scope Owend<ISpan> processingRequestSpan?
+    processingRequestSpan->End();
```
This should be managed by a local class object - otherwise any exceptions thrown will not call this code.

ghalliday commented on esp/bindings/http/platform/httpservice.hpp:
```diff
@@ -34,6 +34,7 @@
 
 #include "espsession.ipp"
 #include "jhash.hpp"
+#include "jtrace.hpp"
```
Should not be needed

ghalliday commented on esp/platform/espp.cpp:
```diff
@@ -483,6 +489,16 @@ int init_main(int argc, const char* argv[])
+        //rootEspProcspan->SetAttribute("application", application);
+
+        std::string traceId;
+        traceManager.getCurrentTraceId(traceId);
```
unused?

ghalliday commented on system/jlib/CMakeLists.txt:
```diff
@@ -230,6 +242,17 @@ if ( ${HAVE_LIBCRYPT} )
+        opentelemetry-cpp::api
+        opentelemetry-cpp::ext
+        opentelemetry-cpp::sdk
+        opentelemetry-cpp::logs #might not need
```
remove them if unlikely to need, and not using at the moment.

ghalliday commented on system/jlib/jtrace.cpp:
```diff
@@ -69,3 +73,190 @@ const char* LogTrace::queryLocalId() const
+    //Target propogator? http? grpc? binary? custom?
+    //HPCC component tracing switches?
+    Owned<const IPropertyTree> traceConfig;
+    try
```
I'm not sure you need the try block.  If this throws an exception it suggests you are calling it at the wrong point.

ghalliday commented on system/jlib/jtrace.hpp:
```diff
@@ -55,6 +231,145 @@ class jlib_decl LogTrace
+static constexpr const char *kCallerIdHTTPHeader = "HPCC-Caller-Id";
+}
+
+// github copilot generated comment:
```
It didn't really make it any clearer!

ghalliday commented on system/jlib/jtrace.hpp:
```diff
@@ -55,6 +231,145 @@ class jlib_decl LogTrace
+
+    TraceManager(const std::string & moduleName)
+    {
+        TraceManager(moduleName.c_str());
```
This is almost certainly going to lead to traceName pointing at a name that has been fred.

ghalliday commented on system/jlib/jtrace.hpp:
```diff
@@ -55,6 +231,145 @@ class jlib_decl LogTrace
+    static opentelemetry::nostd::shared_ptr<opentelemetry::trace::Tracer> getTracer(std::string moduleName);
+
+    //Extracts parent contex from the carrier's headers and returns it as callerSpanId
+    static void getCallerSpanId(context::propagation::TextMapCarrier &carrier, std::string & callerSpanId);
```
I suspect these should be member functions of another class, rather than static functions.

rpastrana replied:
Let's discuss offline to clearly define "another class"

ghalliday commented on system/jlib/jtrace.hpp:
```diff
@@ -55,6 +231,145 @@ class jlib_decl LogTrace
+    // HPCCHttpTextMapCarrier<IProperties> , HttpTextMapCarrier<http_client::Headers>, etc
+    // The carrier must implement the TextMapCarrier interface
+    // and it will determine the header which contains the parent span
+    template<class CARRIER>
```
Again the use of templates suggests a lack of design.   The HPCC codebase tends to use interfaces rather than templates classes. 
templates are generally reserved for 
- simple integer types
- generic general purpose container classes (e.g. hash table)
- rarely implementing classes on top of other base classes.

ghalliday reviewed (commented):
The comments on my first pass are a bit random, but the general impression is:

* The open telemetry code should be encapsulated within jtrace, and ideally would not leak outside it.  (That may not be completely possible, but it is a good initial ideal.)
* The jtrace code should have classes which map to the different concepts that we want to support and can be used to manage them.  That was not clear when I looked at it.
* Generally we use interfaces rather than template classes - see a more detailed comment above.

I will reread it through again before you get back - I expect some of it will become clearer.

afishbeck commented on esp/platform/espp.cpp:
```diff
@@ -426,6 +427,11 @@ int init_main(int argc, const char* argv[])
 
+    TraceManager traceManager("esp");
+    auto espTracer = traceManager.getTracer();//All instances of this esp process can share the same tracer same as TraceManager::getTracer("esp")
+    auto rootEspProcspan = espTracer->StartSpan(__func__); //Dummy top level span for the esp process
```
Is this dummy span really needed?  If so, at least for now, I would completely hide it from a process that handles requests.  Even for job type processes, like say eclagent, I think.

rpastrana replied:
Not required, the component owner will define the appropriate spans, this illustrates a top level span for the current function

afishbeck commented:
I think the general concept is good.  But I would vastly simplify the abstraction, especially in the beginning.  Then as I added features I could be forced to expose a little more complexity. 

But basically my first pass would to implement something like:

```
**process main:**

initTracing("roxie");

**incoming query:**

inside the roxie context, or passed into it:

HttpRequest req;

//the idea of the transaction span is an opentelemetry server span, but I would make a distinction of a process vs a transaction
Owned<ITraceTransactionSpan> tspan = createTracingTransactionSpan(req.headers);  //can grab tracing headers and global-id headers.  Can also grab whatever http attributes we want
tspan->addAttribute(SemanticConventions::kProcessPid, pid); //just an example, not sure which non http semantic conventions we want.
tspan->addAttribute("hpcc.version", ver);  //add well known or custom attributes.  this is where I'd consider not getting in the way of the opentelemetry registered values.  needs some thought.


**ECL code:**

getTraceId(){getRoxieContext->getTransactionSpan()->getTraceId()}
getSpanId(){getRoxieContext->getTransactionSpan()->getSpanId()}
getGlovalId(){getRoxieContext->getTransactionSpan()->getGlobalId()}

//The ECL developers do their own proprietary logging of IDs.

**SOAPCALL:**

Owned<ITraceChildSpan> childspan = getRoxieContext->getTransactionSpan()->createChildSpan("SOAPCALL");

//add well known or custom attributes.  this is where I'd consider not getting in the way of the opentelemetry registered values.  The header with those value names are really just defined strings
span->addAttribute("hpcc.soapcall.timeout", timeout);  


**SocketCall out from soapcall:**

Owned<ITraceClientSpan> clientspan = childspan->createClientSpan("SOAPCALL client");
HttpRequest req;
clientSpan->setHeaders(req.headers, includeTraceHeaders, includeGlobalId);
sendHttpReq(req);

for now, when each span goes out of scope it is logged through jlog, likely in its opentelemetry json format?
Future we can support other exporters.
```

afishbeck commented:
Spans for queued workunits and agent jobs would be a different discussion.

rpastrana commented:
> The comments on my first pass are a bit random, but the general impression is:
> 
>     * The open telemetry code should be encapsulated within jtrace, and ideally would not leak outside it.  (That may not be completely possible, but it is a good initial ideal.)
> 
>     * The jtrace code should have classes which map to the different concepts that we want to support and can be used to manage them.  That was not clear when I looked at it.
> 
>     * Generally we use interfaces rather than template classes - see a more detailed comment above.
> 
> 
> I will reread it through again before you get back - I expect some of it will become clearer.



> 

Great comments and I'm in agreement.

rpastrana committed:
HPCC-29332 WIP

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

rpastrana committed:
HPCC-29332 WIP

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

rpastrana committed:
HPCC-29332 Cleanup

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

ghalliday reviewed (commented):
@rpastrana from an initial scan that looks much closer to what I would hope for.  More of the implementations can be moved from the header file to the cpp file.  I think there is a good chance you can remove all references to open telemetry in the header if you do that.

I will go back and look at it much more closely.

ghalliday commented on system/jlib/jtrace.cpp:
```diff
@@ -69,3 +83,282 @@ const char* LogTrace::queryLocalId() const
+    }
+}
+
+void CTransactionSpan::setAttriburesFromHTTPHeaders(StringArray & httpHeaders)
```
typo: attribu**t**es
same for the function below.

ghalliday commented on system/jlib/jtrace.cpp:
```diff
@@ -69,3 +83,282 @@ const char* LogTrace::queryLocalId() const
+            hpccCallerId.set(httpHeaders.item(currentHeaderIndex+1));
+            continue;
+        }
+    }
```
I would expect this function and the next to do the same thing.  This doesn't contain the latter code.

For simplicitly I would create an IProperties from the array and then call the other function.  If it is time critical then investigate some more later.

ghalliday reviewed (changes requested):
@rpastrana some more detailed comments.  A brief summary form memory:

- Perform the traceManager initialisation once when the config is processed.
- I don't think we need IHPCCTracer as well as ITraceManager
- Simplify the interface a bit by removing a few functions, and not passing attributes when a span is created.
- activate() not needed
- move some more code into the cpp file from the header.

rpastrana committed:
Merge branch 'master' of https://github.com/hpcc-systems/HPCC-Platform into HPCC-29332-OpenTelCpp

rpastrana committed:
HPCC-29332 Interface cleanup

- Removes sample usage to avoid code review complications
- Adds realistic sample in ESP processrequest
- Initialize OpenTel at loadConfig time
- Migrate OT imports to implementation
- Eliminate HPCCTracer interface
- Add appropriate interface methods
- Does not yet use LogTrace

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

ghalliday reviewed (commented):
@rpastrana various comments.  We can walk through them later.

Also you seem to have merged master into this branch - that will cause problems later, you need to rebase onto master instead.

rpastrana committed:
HPCC-29332 Interface cleanup

- Provides simple clear interface definition
- Provides traceman init and query
- Migrate ALL OT imports to implementation
- Add appropriate interface methods
- Provides manual parent span injection
- Does not yet use LogTrace
- Does not yet defint helm/k8s schema

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

rpastrana committed:
HPCC-29332 Code Revie

- Updates interface
- Forces specific Server->client/internal span relationship
- Starts OTLP support
- Attempts to remove all getcurrent references
- Does not yet use LogTrace
- Does not yet defint helm/k8s schema

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

ghalliday reviewed (commented):
@rpastrana a couple of comments, but looks a good base for discussion.

rpastrana commented on system/jlib/jtrace.hpp:
```diff
@@ -55,6 +66,33 @@ class jlib_decl LogTrace
+    virtual const char * queryHPCCCallerID() = 0;
+    virtual const char * querySpanName() = 0;
+    virtual const char * queryTraceID() = 0;
+    virtual const char * queryTraceFlags() = 0;
```
The traceflags are def openTelemetry specific, and I could see an argument to keep the interface agnostic...
They, along with the traced and span id make up the span context which is used to create and inject parentContext

rpastrana commented on system/jlib/jtrace.hpp:
```diff
@@ -55,6 +66,33 @@ class jlib_decl LogTrace
+    virtual void setSpanAttribute(const char * key, const char * val) = 0;
+    virtual void setSpanAttributes(const IProperties * attributes) = 0;
+    virtual void addSpanEvent(const char * eventName) = 0;
+    virtual void setActive() = 0;
```
I debated not exposing this because it is very openTel specific, but ultimately decided the caller should have the ability to set a particular trace as the active (parent span for all new spans). If we don't expose this, or similar functionality the caller will be responsible for tracking and explicitly setting the parent span which is sensible I suppose

ghalliday commented on esp/bindings/http/platform/httpservice.cpp:
```diff
@@ -268,8 +313,16 @@ int CEspHttpServer::processRequest()
         else //user ID is in HTTP header
             ESPLOG(LogMin, "%s %s, from %s@%s", method.str(), m_request->getPath(pathStr).str(), userid, m_request->getPeer(peerStr).str());
-
+//checkUserAuth could declare nested span
```
Should probably also move to the unit tests.  Creating spans when we call ldap etc. is something worth recording in a jira to come back to later.

ghalliday commented on system/jlib/jtrace.hpp:
```diff
@@ -55,6 +66,33 @@ class jlib_decl LogTrace
+
+    virtual void querySpanContextProperties(IProperties * contextProps) = 0;
+    //virtual bool injectSpanContext(CHPCCHttpTextMapCarrier * carrier) = 0;
+    virtual bool injectSpanContext(IProperties * contextProps) = 0;
```
This name could be improved.  Probably get rather than inject.  What is the difference with the function above?  If different I suspect the parameter names should be different, and possibly combine into a single call.

ghalliday commented on system/jlib/jtrace.hpp:
```diff
@@ -55,6 +66,33 @@ class jlib_decl LogTrace
+    virtual void setSpanAttributes(const IProperties * attributes) = 0;
+    virtual void addSpanEvent(const char * eventName) = 0;
+
+    virtual void querySpanContextProperties(IProperties * contextProps) = 0;
```
consistentency: Functions that start query generally return a value.  get is generally used if returns a value in a supplied parameter.

ghalliday commented on system/jlib/jtrace.cpp:
```diff
@@ -69,3 +100,617 @@ const char* LogTrace::queryLocalId() const
+
+            const CHPCCHttpTextMapCarrier carrier(httpHeaders);
+            auto globalPropegator = context::propagation::GlobalTextMapPropagator::GetGlobalPropagator();
+            opentelemetry::v1::context::Context currentContext = context::RuntimeContext::GetCurrent();
```
Calling getCurrent() - I don't think that is valid.

ghalliday commented on system/jlib/jtrace.cpp:
```diff
@@ -69,3 +100,617 @@ const char* LogTrace::queryLocalId() const
+         init();
+    }
+
+    CSpan(const char * spanName, SpanType type, const char * tracerName_)
```
consistency: 3 different ways of differentiating the parameter from the member i) diff name ii) use this for the member iii) append an underscore.
I don't really mind any of them (although everywhere else uses _ as a prefix rather than a suffix), but much better to be consistent.

rpastrana replied:
one of the many reasons I prefer to prefix all members with m_

ghalliday commented on system/jlib/jtrace.cpp:
```diff
@@ -69,3 +100,617 @@ const char* LogTrace::queryLocalId() const
+            tracerName.set(parent->queryTraceName());
+        this->type = type;
+
+         init();
```
indent: extra space

ghalliday reviewed (commented):
@rpastrana a few minor comments.  I think now would be a good time to squash all the commits, close this PR and open a new one with the branch.

rpastrana commented on esp/bindings/http/platform/httpservice.cpp:
```diff
@@ -184,6 +184,48 @@ void checkSetCORSAllowOrigin(EspHttpBinding *binding, CHttpRequest *req, CHttpRe
 int CEspHttpServer::processRequest()
 {
+    //Mock http headers from request
+    Owned<IProperties> mockHTTPHeaders = createProperties();
```
100% agree on the usefulness of unit tests
